Source/DafnyCore/Auditor/Assumption.cs

```diff
@@ -0,0 +1,211 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Dafny.Auditor;
+
+/// <summary>
+/// A set of assumption-related properties of a Dafny entity. Certain
+/// combinations of these properties classify the entity as containing
+/// an assumption.
+/// </summary>
+public struct AssumptionProperties {
+  internal bool IsGhost;
+  internal bool IsSubsetType;
+  internal bool IsCallable;
+  internal bool IsTrait;
+  internal bool MissingBody;
+  internal bool HasAxiomAttribute;
+  internal bool HasExternAttribute;
+  internal bool HasVerifyFalseAttribute;
+  internal bool HasAssumeInBody;
+  internal bool MayNotTerminate;
+  internal IEnumerable<AttributedExpression> RequiresClauses; // non-null implies one or more elements
+  internal IEnumerable<AttributedExpression> EnsuresClauses; // non-null implies one or more elements
+
+  /// <summary>
+  /// Determines whether the set of properties constitutes an assumption.
+  /// </summary>
+  /// <returns>whether this entity contains assumptions</returns>
+  public bool HasAssumptions() {
+    return HasAxiomAttribute ||
+           (IsCallable &&
+            (((EnsuresClauses is not null || RequiresClauses is not null) &&
+              (MissingBody || HasExternAttribute)) ||
+             HasAssumeInBody)) ||
+           (MayNotTerminate && (IsCallable || IsTrait));
+  }
+
+  /// <summary>
+  /// Describe all of the assumptions this entity makes.
+  /// </summary>
+  /// <returns>a list of (assumption, mitigation) pairs</returns>
+  public IEnumerable<(string, string)> Description() {
+    if (HasAxiomAttribute) {
+      return new List<(string, string)>{
+        ("Declaration has explicit `{:axiom}` attribute.",
+         "Provide a proof, test, or other justification.")
+      };
+    }
+
+    List<(string, string)> descs = new();
+
+    if (IsCallable && MissingBody) {
+      descs.Add(
+        ((IsGhost ? "Ghost" : "Compiled") +
+          " declaration has no body" +
+          (EnsuresClauses is null ? "." : " and has an ensures clause."),
+        "Provide a body or add `{:axiom}`."));
+    }
+    if (IsCallable && HasExternAttribute && RequiresClauses is not null) {
+      descs.Add(
+        ("Declaration with `{:extern}` has a requires clause.",
+         "Test external caller (maybe with `/testContracts`) or justify."));
```

Comment:
I might remove the `justify` text, or enable it optionally. The reason for it is that output produced in the `markdown-ietf` format can be used by [`Duvet`](https://github.com/awslabs/duvet), which can then keep track of whether there are comments in the source code explaining why a particular assumption has been made. In this particular case, one of those comments would probably appear on the code that's responsible for testing the contract on the extern. However, it doesn't make much sense when not using that output format.

As for `requires` clauses, it depends on how the `{:extern}` is being used. It can be used to mark code as being called by an external client (and possibly to change the compiled name in that case). If it's being used that way, then any `requires` clause is an unchecked assumption. Ideally, we would distinguish between external "imports" and "exports", and warn about `ensures` clauses in the former case and `requires` clauses in the latter case. That'll have to wait for the FFI redesign that @robin-aws is working on though.

Reply:
Can you explain what actions users can take to remove flagged code from the report? If there are situations where that is not possible, then I don't think the report will be effective.

Reply:
> As for requires clauses, it depends on how the {:extern} is being used. It can be used to mark code as being called by an external client (and possibly to change the compiled name in that case). If it's being used that way, then any requires clause is an unchecked assumption

Oh yeah indeed that makes sense. I hadn't thought of that. It's a way to fail more gracefully. How would could you provide the information that the external caller's interface is tested?

Reply:
If you use Duvet, you can add a comment to the code doing the testing that declares that it tests that property. There isn't immediately a way to do that with the other report formats.